### PR TITLE
[JENKINS-53250] Gerrit Trigger plugin compatibility with JCasC (JEP-201)

### PR DESCRIPTION
**Pull request by olivergondza**
Add support for injecting configuration through Configuration as Code.

### CODE REVIEW

File: pom.xml
Comment (olivergondza) [inline pom.xml:242]: Majority of the changes to pom ware required by maven enforcer.

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/VerdictCategory.java
Comment (olivergondza) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/VerdictCategory.java:44]: Renamed to match getters so CasC recognize them as properties.

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java
Comment (olivergondza) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java:869]: These methods ware renamed to match respective getters. Otherwise, CasC do not recognize them as a getter/setter for the same property.

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/WorkflowTest.java
Comment (olivergondza) [inline src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/WorkflowTest.java:252]: s/`isA`/`instanceOf`/ required by hamcrest update.

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/GerritJcascConfigurator.java
Comment (rsandell) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/GerritJcascConfigurator.java:111]: This will sort of leave the configuration in a bad state. The verification should be done before setting the new values, somehow...

Comment (rsandell) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/GerritJcascConfigurator.java:111]: This will sort of leave the configuration in a bad state. The verification should be done before setting the new values, somehow...

**Review by rsandell — CHANGES_REQUESTED** [review]

File: pom.xml
Comment (olivergondza) [reply pom.xml:267]: Potentially, but we would have depend on conditional loading of `GerritJcascConfigurator`, which I consider less desirable than adding this potentially unused dependency. Also, I doubt any instance will be prone of configuration-as-code plugin nowadays (when every other plugin depend on it).

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java
Comment (olivergondza) [reply src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java:1012]: Sorry, I am not sure I am following you here. Is that feature about to be reworked/removed? If so, I am ok to drop that from JcasC support in current form (as the change would also affect JCasC DSL).

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/GerritJcascConfigurator.java
Comment (olivergondza) [reply src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/GerritJcascConfigurator.java:111]: Hmm, good point.

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java
Comment (rsandell) [reply src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java:1012]: Yes, the build*values are being worked on to be made dynamically configurable, no longer hard coded to "code review" and "verified". I don't know how the progress is on that effort though.

File: pom.xml
Comment (rsandell) [reply pom.xml:267]: When I've done jcasc compatibility work in other plugins I've made an effort to make it optional ;)

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java
Comment (olivergondza) [reply src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java:1012]: Ok, so I will not expose this through JCasC in this PR not to expose this before rework. It can be always done afterwards.

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java
Comment (olivergondza) [reply src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java:880]: Done

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/GerritJcascConfigurator.java
Comment (olivergondza) [reply src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/GerritJcascConfigurator.java]: Moved.

File: pom.xml
Comment (olivergondza) [reply pom.xml:268]: Done

**Comment by olivergondza** [comment]
@rsandell, all comments addressed, please re-review.

**Comment by olivergondza** [comment]
@rsandell, can you please take another look? Thanks!

**Comment by pjanouse** [comment]
@rsandell can you take a look and if all previous concerns are resolved then go forward and make a release soon please?

We have couple of users waiting on a new release with this changes incorporated.

Thank you!

**Comment by sciyako** [comment]
We are eager for this to be released as well, one of the big blocking points for our deployment.

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/PluginImpl.java
Comment (lucamilanesio) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/PluginImpl.java:429]: Same as above: why this setter needs to be synchronized?

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java
Comment (lucamilanesio) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java:626]: Should the old setting be flagged as `deprecated`?

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java
Comment (lucamilanesio) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java:856]:
```suggestion
     * @param buildCurrentPatchesOnly new policy.
```

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java
Comment (lucamilanesio) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java:854]:
```suggestion
     * Set build cancellation policy for current patches
     *
     * Exposed for configuration as code.
```

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java
Comment (lucamilanesio) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java:879]:
```suggestion
     * Set Gerrit command for successful verification.
```

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java
Comment (lucamilanesio) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java:881]:
```suggestion
     * @param cmd the Gerrit command for successful verification
```

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java
Comment (lucamilanesio) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java:905]:
```suggestion
     * Set Gerrit command for unstable builds
```

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java
Comment (lucamilanesio) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java:1260]: Nit: avoid mixing formatting and functional changes.

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/GerritJcascConfigurator.java
Comment (lucamilanesio) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/GerritJcascConfigurator.java:281]: Have you checked the correct indenting here?

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/mock/MockGerritHudsonTriggerConfig.java
Comment (lucamilanesio) [inline src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/mock/MockGerritHudsonTriggerConfig.java:464]: nit: separate formatting from functional changes

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/PluginImpl.java
Comment (olivergondza) [reply src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/PluginImpl.java:420]: Yes it is related. JCasC thread is calling this setter from separate thread and this to make sure the change propagates.

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java
Comment (olivergondza) [reply src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java:626]: I had no clue this option is deprecated. I am ok to leave it out from JCasC support, if that makes sense. Exposing the option through JCasC as deprecated and subject of eventual removal makes less sense to me.

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/GerritJcascConfigurator.java
Comment (olivergondza) [reply src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/GerritJcascConfigurator.java:281]: The codebase does not seem to be consistent on indenting following lines of a single expression broken over multiple lines. I do prefer indenting such lines by a multiple of normal indent (`4` vs `8`) in this case. `PluginConfig#setValues` alone mixes both stiles. :shrugs:

**Comment by olivergondza** [comment]
@lucamilanesio, thanks for your review. PR amended.

Is there anything else that needs to be done before we can merge this? @rsandell?

**Comment by lucamilanesio** [comment]
> @lucamilanesio, thanks for your review. PR amended.

Thanks @olivergondza for your amendements.
There is a checkstyle issue to fix I believe.

**Comment by olivergondza** [comment]
> There is a checkstyle issue to fix I believe.

Yep, that is a result of removing the unrelated change in `MockGerritHudsonTriggerConfig.java`: https://github.com/jenkinsci/gerrit-trigger-plugin/pull/426/commits/59940315412ab8c8b3ef98480164c66553647631#diff-927a2e97d7dbc0d05f131db82f904ed12244e4b64ba39ac46e1d86dd2590004bR462

**Comment by sciyako** [comment]
It's been a few days now with all checks passed, could we get this merged please ?

**Review by lucamilanesio — APPROVED** [review]
@rsandell would you like to double-check if you have anything else to report?

**Comment by lubomir** [comment]
I'd really love to see this available in a released version. Any estimate on when it can be merged?

**Comment by MosheSheena** [comment]
Joining in on the request-merge party here - we really need this to be merged as soon as possible.
Thanks!

**Comment by lybaokhanh** [comment]
+1 member on request-merge party, love to hear the good news soon.

**Review by rsandell — APPROVED** [review]

**Comment by rsandell** [comment]
@olivergondza I was on my way of releasing this on Friday, but noticed after the merge that the version of jackson has known CVEs reported on it. So I started the work on migrating to use `org.jenkins-ci.plugins:jackson2-api` but that led me down the rabbit hole of fixing enforcement rules and mocked tests that needed adapting to the new core etc. etc. etc. So I never managed to release it.
And now I don't know when I will get time to work on that again :cry:

**Comment by olivergondza** [comment]
@rsandell, it is likely desirable to move to the API plugin, but it seems enough for now to advance the jackson dependency. Let me have a look at that

**Comment by olivergondza** [comment]
@rsandell, actually https://github.com/jenkinsci/gerrit-trigger-plugin/pull/434 applies cleanly and it resolves the issue IMO.

**Comment by rsandell** [comment]
@olivergondza OK, that can work for now I guess :)

File: pom.xml
Comment (basil) [inline pom.xml:281]: Causes [JENKINS-65650](https://issues.jenkins.io/browse/JENKINS-65650).

Comment (fraz3alpha) [reply pom.xml:281]: Also causes https://issues.jenkins.io/browse/JENKINS-65771 :(

Comment (olivergondza) [reply pom.xml:281]: What a rookie mistake. Resoled in alpha in https://github.com/jenkinsci/gerrit-trigger-plugin/commit/63407f4c4a314043b39b21c5cc340bb9a8200e7b

Comment (olivergondza) [reply pom.xml:281]: @rsandell, Can you please cut 2.35 (there is only a beta out). It affects unrelated plugins as Jenkins introspect this test-only dependency. Thanks!

**Pull request merged**
